// fumubang/index.js
/**
 * Created by zzy on 2014/9/29.
 */
var request = require('request');
var async = require('async');
var cheerio = require('cheerio');
var _ = require('underscore');
var trader = '5434d488a1efc590111dd8f1';
var j = request.jar();
var timeZone = ' 00:00:00 +08:00';
request = request.defaults({jar:j});
var logined = false;
//'login':'太仓盛兴','pass':'sty841'

var FMB = function(name,passwd){
    this.name = name;
    this.passwd = passwd;

};

FMB.prototype.login = function(fn){
    request.post('http://www.fumubang.com/shoplogin/dologin',{'form':{'login':this.name,'pass':this.passwd}}, function optionalCallback(err, httpResponse, body) {
        if(!err&&body=='登陆成功'){
            logined = true;
            fn(null,true);
        } else {
            fn(new Error('登陆失败'),false);
        }
    });
}

FMB.prototype.verifyCode = function(code,fn){
    var that = this;
    async.series([
        function(cb){
            if(!logined){
                that.login(function(err,res){
                    cb(err,res);
                });
            } else {
                cb(null,null);
            }
        },
        function(cb){
            request.post('http://www.fumubang.com/shoplogin/operCode',{'form':{'exchange_code[]':code}}, function optionalCallback(err, httpResponse, body) {
                if(err){
                    cb(err,null);
                } else {
                    var result = JSON.parse(body).data[0];
                    if(result.state==0){
                        cb(new Error(result.error),null);
                    } else {
                        cb(null,result.data[0]);
                    }
                }
            })
        }
    ],function(err,res){
       fn(err,res[1]);
    });
};

FMB.prototype.getOrders = function(){
    async.auto({
        'getToken':function(cb){
            request.get('http://172.16.0.15:3000/api/member/login?username=fmb_system&passwd=e10adc3949ba59abbe56e057f20f883e',
                function optionalCallback(err, httpResponse, body) {
                    if (err) {
                        cb(err,null);
                    } else {
                        var result = JSON.parse(body);
                        console.log(typeof(result.error),result.error);
                        if(result.error!=0){
                            console.log('1');
                            cb(new Error(result.errMsg),null);
                        } else {
                            console.log('0');
                            cb(null,result.data.token);
                        }
                    }
                });
        },
        'getTotalSize':function(cb){
            request.get('http://www.fumubang.com/shoplogin/orderCat/', function optionalCallback(err, httpResponse, body) {
                if (err) {
                    cb(err,null);
                } else {
                    var $ = cheerio.load(body);
                    var re=/\d+/;
                    var totalSize = parseInt($('.ord_result h1').text().match(re)[0]);
                    cb(null,totalSize)
                }
            });
        },
        'getOrders':['getTotalSize',function(cb,results){
            getPageOrder(results.getTotalSize,30,function(err,res){
                cb(err, _.flatten(res));
            });
        }],
        'saveOrder':['getOrders','getToken',function(cb,results) {
            var orders = results.getOrders;
            console.log(orders[0]);
            request.post('http://172.16.0.15:3000/api/order/traderOrder',
                {'form': {
                    'trader': trader,
                    'token': results.getToken,
                    'startDate': new Date(orders[0].startDate+timeZone).getTime(),
                    'quantity': orders[0].qty,
                    'remark': '',
                    'traderProduct': orders[0].productId
                }},
                function optionalCallback(err, httpResponse, body) {
                    if (err) {
                        cb(err, null);
                    } else {
                        cb(null, JSON.parse(body));
                    }
                })
        }]
    },function(err,res){
        console.log(err,res.saveOrder);
    });
};

function getPageFun(i){
    return function(cb){
        request.get('http://www.fumubang.com/shoplogin/orderCat/?page='+i, function optionalCallback(err, httpResponse, body) {
            if (err) {
                cb(err,null);
            } else {
                var $ = cheerio.load(body);
                var orders = [];
                var order;
                $('.orderck-dl').children().each(function(i,e){
                    if(i!=0){
                        if(i%2!=0){
                            order = {};
                            order.orderId = $(e).find('a').text();
                            order.orderDate = $(e).find('span').eq(0).text();
                            order.payDate =  $(e).find('span').eq(1).text();
                        } else {
                            order.name = $(e).find('a').eq(0).text();
                            order.productId = $(e).find('a').eq(0).attr('href').replace('/','').replace('.html','');
                            order.sellPrice = $(e).find('td').eq(1).text();
                            order.settlementPrice = $(e).find('td').eq(2).text();
                            order.qty = $(e).find('td').eq(3).text();
                            order.refundStatus = $(e).find('td').eq(4).text();
                            order.totalPrice = $(e).find('td').eq(5).text();
                            order.stauts = $(e).find('td').eq(6).text();
                            order.startDate = $(e).find('p').eq(1).text().match( /([1-2]\d{3})[\/|\-](0?[1-9]|10|11|12)[\/|\-]([1-2]?[0-9]|0[1-9]|30|31)/)[0];
                            orders.push(order);
                        }
                    }
                });
                cb(null,orders);
            }
        });
    }
}

function getPageOrder(totalSize,pageSize,fn){
    var totalPage = Math.ceil(totalSize/pageSize);
    var funcArray = [];
    for(var i=1;i<=totalPage;i++){
        funcArray.push(getPageFun(i));
    }
    async.series(funcArray,function(err,res){
        fn(err,res);
    })
}

module.exports=FMB;

//async.auto({
//    'getToken':function(cb){
//        request.get('http://172.16.0.15:3000/api/member/login?username=fmb_system&passwd=e10adc3949ba59abbe56e057f20f883e',
//            function optionalCallback(err, httpResponse, body) {
//                if (err) {
//                    cb(err,null);
//                } else {
//                    var result = JSON.parse(body);
//                    console.log(typeof(result.error),result.error);
//                    if(result.error!=0){
//                        console.log('1');
//                        cb(new Error(result.errMsg),null);
//                    } else {
//                        console.log('0');
//                        cb(null,result.data.token);
//                    }
//                }
//            });
//    },
//    'login':function(cb){
//        request.post('http://www.fumubang.com/shoplogin/dologin',{'form':{'login':'太仓盛兴','pass':'sty841'}}, function optionalCallback(err, httpResponse, body) {
//            if (err) {
//                cb(err,null);
//            } else {
//                cb(null,null)
//            }
//        })
//    },
//    'getTotalSize':['login',function(cb){
//        request.get('http://www.fumubang.com/shoplogin/orderCat/', function optionalCallback(err, httpResponse, body) {
//            if (err) {
//                cb(err,null);
//            } else {
//                var $ = cheerio.load(body);
//                var re=/\d+/;
//                var totalSize = parseInt($('.ord_result h1').text().match(re)[0]);
//                cb(null,totalSize)
//            }
//        });
//    }],
//    'getOrders':['getTotalSize',function(cb,results){
//        getPageOrder(results.getTotalSize,30,function(err,res){
//            cb(err, _.flatten(res));
//        });
//    }],
//    'saveOrder':['getOrders','getToken',function(cb,results) {
//        var orders = results.getOrders;
//        console.log(orders[0]);
//        request.post('http://172.16.0.15:3000/api/order/traderOrder',
//            {'form': {
//                'trader': trader,
//                'token': results.getToken,
//                'startDate': new Date(orders[0].startDate+timeZone).getTime(),
//                'quantity': orders[0].qty,
//                'remark': '',
//                'traderProduct': orders[0].productId
//            }},
//            function optionalCallback(err, httpResponse, body) {
//                if (err) {
//                    cb(err, null);
//                } else {
//                    cb(null, JSON.parse(body));
//                }
//            })
//    }]
//},function(err,res){
//
//    console.log(err,res.saveOrder);
//});
//async.waterfall([
//    //登陆
//    function(cb){
//        request.post('http://www.fumubang.com/shoplogin/dologin',{'form':{'login':'太仓盛兴','pass':'sty841'}}, function optionalCallback(err, httpResponse, body) {
//            if (err) {
//                cb(err,null);
//            } else {
//                cb(null,httpResponse.headers['set-cookie'])
//            }
//        });
//    },
//    //获取总条数
//    function(cookies,cb){
//        request.get('http://www.fumubang.com/shoplogin/orderCat/', function optionalCallback(err, httpResponse, body) {
//            if (err) {
//                cb(err,null);
//            } else {
//                var $ = cheerio.load(body);
//                var re=/\d+/;
//                var totalSize = parseInt($('.ord_result h1').text().match(re)[0]);
//                cb(null,totalSize)
//            }
//        });
//    },
//    //获取订单
//    function(totalSize,cb){
//        getPageOrder(totalSize,30,function(err,res){
//            cb(err, _.flatten(res));
//        });
//    },
//    //保存订单
//    function(orders,cb){
//        /*
//        * var trader = request.body.trader;
//         var token = request.body.token;
//         var startDate = request.body.startDate;
//         var quantity = request.body.quantity;
//         var remark = request.body.remark;
//         var traderProduct = request.body.traderProduct;
//         var liveName = request.body.liveName;
//         var contactPhone = request.body.contactPhone;
//        * */
//        request.post('http://172.16.0.15:3000/api/order/traderOrder',
//            {'form':{
//                'trader':trader,
//                'token':'token',
//                'startDate':new Date(orders[0].startDate),
//                'quantity':orders[0].qty,
//                'remark':'',
//                'traderProduct':orders[0].productId
//            }},
//            function optionalCallback(err, httpResponse, body) {
//            if (err) {
//                cb(err,null);
//            } else {
//                cb(null,httpResponse.headers['set-cookie'])
//            }
//        });
//    }
//],function(err,res){
//    console.log(err,res,res.length);
//})
